### PR TITLE
Improved performance with large HTML files

### PR DESCRIPTION
**Pull request by sapphi-red**
## Description

I reduced copying AST like #11108. 

On [this case](https://github.com/prettier/prettier/issues/4776#issuecomment-740791694), the result of `yarn run perf:benchmark` became about 8.6x faster.

before
```
[debug] resolve config from 'test.html'
[debug] loaded options `{"useTabs":false,"tabWidth":2,"endOfLine":"lf"}`
[debug] applied config-precedence (cli-override): {"endOfLine":"lf","tabWidth":2,"useTabs":false}
[debug] '--debug-benchmark' option found, measuring formatWithCursor with 'benchmark' module.
[debug] '--debug-benchmark' measurements for formatWithCursor: {
[debug]   "benchmark": "format x 0.80 ops/sec ±3.60% (7 runs sampled)",
[debug]   "hz": 0.8011225145560241,
[debug]   "ms": 1248.2485285714286
[debug] }
```

after
```
[debug] resolve config from 'test.html'
[debug] loaded options `{"useTabs":false,"tabWidth":2,"endOfLine":"lf"}`
[debug] applied config-precedence (cli-override): {"endOfLine":"lf","tabWidth":2,"useTabs":false}
[debug] '--debug-benchmark' option found, measuring formatWithCursor with 'benchmark' module.
[debug] '--debug-benchmark' measurements for formatWithCursor: {
[debug]   "benchmark": "format x 6.86 ops/sec ±11.89% (23 runs sampled)",
[debug]   "hz": 6.86268975187975,
[debug]   "ms": 145.71546086956522
[debug] }
```

refs #4776 

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Thanks for working on this.

>  the result of yarn run perf:benchmark became about 8.6x faster.

👍

**Comment by sapphi-red** [comment]
@fisker Is there anything I should do for the next step?

File: src/language-html/ast.js [hunk 1]
```diff
@@ -84,6 +84,68 @@ class Node {
...
+      parent: this,
+    });
+
+    let next = target;
+    while (next) {
+      // @ts-expect-error
+      next.index++;
+      // @ts-expect-error
+      next = next.next;
+    }
```
Comment (fisker) [inline src/language-html/ast.js:108]: Is  it possible make this run into an infinite loop? Maybe we can recalculate `index`/`next` by looping children?

**Comment by sapphi-red** [comment]
I have changed the code to use for loop to recalculate `index`/`next`/`prev` as you suggested, which would be much more readable!
The benchmark result for this code is below.
```
[debug] resolve config from 'test.html'
[debug] loaded options `{"useTabs":false,"tabWidth":2,"endOfLine":"lf"}`
[debug] applied config-precedence (cli-override): {"endOfLine":"lf","tabWidth":2,"useTabs":false}
[debug] '--debug-benchmark' option found, measuring formatWithCursor with 'benchmark' module.
[debug] '--debug-benchmark' measurements for formatWithCursor: {
[debug]   "benchmark": "format x 5.21 ops/sec ±10.93% (19 runs sampled)",
[debug]   "hz": 5.207000038723637,
[debug]   "ms": 192.04916315789475
[debug] }
```

File: src/language-html/ast.js [hunk 2]
```diff
@@ -84,6 +84,42 @@ class Node {
...
+  _updateChildrenWithoutClone() {
+    // @ts-expect-error
+    for (let i = 0; i < this.children.length; i++) {
+      // @ts-expect-error
+      setNonEnumerableProperties(this.children[i], {
+        index: i,
+        // @ts-expect-error
+        prev: i > 0 ? this.children[i-1] : undefined,
+        // @ts-expect-error
+        next: i < this.children.length - 1 ? this.children[i+1] : undefined,
```
Comment (fisker) [inline src/language-html/ast.js:117]:
```suggestion
        // @ts-expect-error
        prev: this.children[i-1],
        // @ts-expect-error
        next: this.children[i+1],
```

File: src/language-html/ast.js [hunk 3]
```diff
@@ -84,6 +84,42 @@ class Node {
...
+  /**
+   * @param {Node} [child]
+   */
+  removeChild(child) {
+    // @ts-expect-error
+    this.children.splice(this.children.indexOf(child), 1);
+    this._updateChildrenWithoutClone();
+  }
+
+  _updateChildrenWithoutClone() {
```
Comment (fisker) [inline src/language-html/ast.js:108]: Shouldn't it enough only fix `self`, `prev`, `next`? We already know the index.

File: src/language-html/parser-html.js [hunk 4]
```diff
@@ -346,19 +349,20 @@ function _parse(text, options, parserOptions, shouldParseFrontMatter = true) {
...
+  ast.walk((node) => {
     if (node.type === "comment") {
       const ieConditionalComment = parseIeConditionalComment(
         node,
         parseSubHtml
       );
       if (ieConditionalComment) {
-        return ieConditionalComment;
+        node.parent.insertChildBefore(node, ieConditionalComment);
+        node.parent.removeChild(node);
```
Comment (fisker) [inline src/language-html/parser-html.js:360]: A new `.replaceChild()` method?

**Comment by fisker** [comment]
I'm not sure about the performance impact, it seems we can use `NODES_KEYS` as a `Set` instead of an object. There are lots of `for..in` loop.

**Comment by fisker** [comment]
I'm not familiar with this part of code. Why are we using descriptors instead of class getter?

**Comment by sapphi-red** [comment]
> I'm not sure about the performance impact, it seems we can use `NODES_KEYS` as a `Set` instead of an object. There are lots of `for..in` loop.

I tried this now but there was no significant difference (using `Set` was only about 1ms slower).

**Comment by sapphi-red** [comment]
> I'm not familiar with this part of code. Why are we using descriptors instead of class getters?

I'm not sure.
I thought there is no big reason so I rewrote it to class getters and properties.

|commit||||
|---|---|---|---|
| 008975f5f | 136.69717916666664ms | x 7.32 ops/sec | ±13.22% |
| 00099cab9 | 135.36796250000003ms | x 7.39 ops/sec | ±11.86% |
| f6096d2f9 | 142.9828695652174ms | x 6.99 ops/sec | ±11.72% |
| eb0b59f7e | 141.92638695652172ms | x 7.05 ops/sec | ±11.45% |
| e9adae2fc | 160.52020476190478ms | x 6.23 ops/sec | ±11.29% |
| 381a73c16 | 175.88805ms | x 5.69 ops/sec | ±10.47% |

File: src/language-html/print-preprocess.js [hunk 5]
```diff
@@ -122,42 +105,31 @@ function mergeIeConditonalStartEndCommentIntoElementOpeningTag(
...
-              ),
-            })
-          );
+        if (child.type !== "text") {
+          child.type = "text";
+          child.value = getValue(child);
+        }
+
+        const prevChild = node.children[i - 1];
+        if (!prevChild || prevChild.type !== "text") {
```
Comment (fisker) [inline src/language-html/print-preprocess.js:121]: Seems we can use `child.prev` here.

File: src/language-html/print-preprocess.js [hunk 6]
```diff
@@ -190,36 +162,30 @@ function mergeSimpleElementIntoText(ast /*, options */) {
...
-          }
+      for (let i = 0; i < node.children.length; i++) {
+        const child = node.children[i];
+        if (!isSimpleElement(child)) {
+          continue;
         }
-        node.setChildren(newChildren);
+
+        const prevChild = node.children[i - 1];
+        const nextChild = node.children[++i];
```
Comment (fisker) [inline src/language-html/print-preprocess.js:172]: `node.prev` & `node.next` ?

File: src/language-html/print-preprocess.js [hunk 7]
```diff
@@ -319,94 +281,91 @@ function extractWhitespaces(ast /*, options*/) {
...
+        }
 
-          const localChildren = [];
+        const { leadingWhitespace, text, trailingWhitespace } =
+          getLeadingAndTrailingHtmlWhitespace(child.value);
 
-          const { leadingWhitespace, text, trailingWhitespace } =
-            getLeadingAndTrailingHtmlWhitespace(child.value);
+        const prevChild = node.children[i - 1];
+        const nextChild = node.children[i + 1];
```
Comment (fisker) [inline src/language-html/print-preprocess.js:295]: same  here

File: src/language-html/ast.js [hunk 8]
```diff
@@ -85,10 +82,40 @@ class Node {
...
+
+  /**
+   * @param {Node} [target]
+   * @param {Object} [node]
+   */
+  replaceChild(target, node) {
+    const newNode = new Node(node, this);
+
+    // @ts-expect-error
+    this.children.splice(this.children.indexOf(target), 1, newNode);
```
Comment (fisker) [inline src/language-html/ast.js:111]:
```suggestion
    this.children[this.children.indexOf(target)] = newNode;
```

File: src/language-html/ast.js [hunk 9]
```diff
@@ -12,25 +12,31 @@ const NODES_KEYS = {
 // https://github.com/microsoft/TypeScript/issues/26811
 
 class Node {
-  constructor(props = {}) {
+  constructor(props = {}, parent) {
```
Comment (fisker) [inline src/language-html/ast.js:15]: This `parent` argument is weird to me, especially when use `node.clone(parent)`, maybe `new Node({..., parent})`?

**Review by fisker — APPROVED** [review]
LGTM, thank you!

**Review by duailibe — APPROVED** [review]

**Pull request merged**